### ISSUE TITLE
Unexpected behavior regarding `attrs` support

### ISSUE THREAD

**Issue by sasanjac** [snok/flake8-type-checking#79]
I'm really sorry, but I think you may need to revert my changes regarding `attrs` support. It seems that `attrs`  doesn't necessarily need runtime type annotiations and moving all out of `TYPE_CHECKING` block may result in circular imports.

I will investigate how `attrs`  handles runtime type annotations and come back with a fixed proposal.

**Comment by sondrelg** [comment]
Then perhaps instead of reverting the PR, we could add a `type-checking-cattrs-enabled` setting which marks `attrs` type hints as needed at runtime when true? Would the decorators we've used still be exactly the same?

For some context: in the pydantic case we cannot know which classes are pydantic classes, so the behavior for that feature (regrettably) is to just always allow class var type hints as long as the flag is true. Having a `cattrs` settings which opts into the code you've already written already sounds much more robust and precise than that.

**Comment by sasanjac** [comment]
In principle that should work, however, having no imports in `TYPE_CHECKING` blocks may lead to circular import errors. I don't know how `pydantic` handles that since it actually needs the type defs at runtime.

But I think your idea is fine and maybe when the `cattrs` devs introduce some kind of mechanism to mark `cattrs` classes, it can be changed to that marker. The underlying routines should exactly the ones that I implemented, just checking against a different decorator then.

**Comment by sondrelg** [comment]
Would it not simply require

1. A setting defined in the plugin, and passed to the `TypingOnlyImportsChecker`
2. An if-statement to only run the attrs code if the setting is true?

**Comment by sasanjac** [comment]
Yes, but this approach will include all `attrs` classes in a package. With an additional marker you could specify exactly which classes will be part of a `cattrs` structure call. Sadly you can't check against a `cattrs` import, see the following example:

`main.py`
```
import cattrs
import json
from . import Topology

with open("data/topology.json") as f:
    data = json.load(f)

topology = cattrs.structure(Topology, data)

```
`topology.py`
```
import attrs
from . import Node

@attrs.define
class Topology:
    node: Node
```
`node.py`
```
import attrs
import Array

@attrs.define
class Node:
    value: Array
```
`other_node.py`
```
import attrs
import AnotherArray

@attrs.define
class OtherNode:
    value: AnotherArray
```

Here, `AnotherArray` should be put into a `TYPE_CHECKING` block because `OtherNode` is not used in any `cattrs` `structure` call, however, since all `attrs` class attributes are exempt, the plugin won't tell us.

But I think it would be a good enough implementation for now.

**Comment by sondrelg** [comment]
It seems that the `cattrs` case extends to other types of classes, not just `attrs`. For example, this will fail:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

import cattr

if TYPE_CHECKING:
    from typing import List


@dataclass
class S:
    s: List[int]


cattr.unstructure(S([1, 2, 3]))
```

But maybe we can just add dataclass-detection to the cattrs setting

**Comment by sondrelg** [comment]
Think we can consider this closed. Do you agree @sasanjac? 🙂
